common/src/main/java/org/conscrypt/OpenSSLXDHKeyFactory.java

```diff
@@ -47,6 +48,10 @@ protected PublicKey engineGeneratePublic(KeySpec keySpec) throws InvalidKeySpecE
         if (keySpec instanceof X509EncodedKeySpec) {
             return new OpenSSLX25519PublicKey((X509EncodedKeySpec) keySpec);
         }
+        if (keySpec instanceof SecretKeySpec
+            && "RAW".equalsIgnoreCase(((SecretKeySpec) keySpec).getAlgorithm())) {
+            return new OpenSSLX25519PublicKey(((SecretKeySpec) keySpec).getEncoded());
```

Comment:
I don't know the context here, but how come a *Secret*KeySpec is allowed to be an OpenSSLX25519*Public*Key?

Reply:
I'll defer to @prbprbprb here, but we did this for back compat with older android versions.

Reply:
If we expose a new class `XdhKeySpec` , how will android get access to this? Would we introduce a new class on android as well which will delegate to this one? @prbprbprb

Reply:
Fair point, that's going to be a little fugly.

How we dealt with this for Scrypt, where there were no standard API classes and we wanted to make it inter-work with Bouncy Castle, we treated the KeySpec as a Java Bean and so anything with the right fields (e.g. an Android public class) would work: https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ScryptSecretKeyFactory.java

Reply:
Gotya, I've created a XdhKeySpec with just one field, the key, which could be either a public or private. The key length check will be added on a separate PR.